src/main/kotlin/org/cascadebot/cascadebot/utils/placeholders/Placeholders.kt

```diff
@@ -0,0 +1,164 @@
+package org.cascadebot.cascadebot.utils.placeholders
+
+import net.dv8tion.jda.api.events.guild.member.GenericGuildMemberEvent
+import net.dv8tion.jda.api.events.guild.member.GuildMemberJoinEvent
+import net.dv8tion.jda.api.events.guild.member.GuildMemberLeaveEvent
+import org.cascadebot.cascadebot.commandmeta.CommandContext
+import org.cascadebot.cascadebot.data.language.Language
+import org.cascadebot.cascadebot.data.language.Locale
+import org.cascadebot.cascadebot.utils.FormatUtils
+import java.time.Duration
+import java.time.OffsetDateTime
+
+class Placeholders<T>(var key: String) {
+
+    companion object {
+        // https://regex101.com/r/LxMI0R/3
+        val placeholderRegex = Regex("\\{([\\p{Ll}_-]+)(?::((?:[\\p{Ll}0-9_-]+,?)+))?}")
+    }
+
+    private val _placeholders: MutableList<Placeholder<T>> = mutableListOf()
+    val placeholders: List<Placeholder<T>>
+        get() = _placeholders.toList()
+
+    fun staticPlaceholder(key: String, mapping: StaticPlaceholder<T>.(PlaceholderContext<T>) -> String?) {
+        _placeholders.add(StaticPlaceholder(key, "${this.key}.$key", mapping))
+    }
+
+    fun argsPlaceholder(key: String, mapping: ArgsPlaceholder<T>.(PlaceholderContext<T>, List<String>) -> String?) {
+        _placeholders.add(ArgsPlaceholder(key,"${this.key}.$key", mapping))
+    }
+
+    // TODO: Use locale!
```

Comment:
I would like to see this done before this gets merged

Reply:
I think it already done, I just need to remove the TODO

Reply:
ok